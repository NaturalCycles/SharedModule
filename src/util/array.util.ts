export function flatten<T> (arrays: T[][]): T[] {
  // to flat single level array
  return ([] as T[]).concat(...arrays)
}
