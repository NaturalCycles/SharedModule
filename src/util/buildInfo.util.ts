import { BuildInfo } from './buildInfo.model'
import {
  gitCurrentBranchName,
  gitCurrentCommitSha,
  gitCurrentCommitTimestamp,
  gitCurrentRepoName,
} from './git.util'

export async function generateBuildInfo (dev = false): Promise<BuildInfo> {
  const [rev, branchName, repoName, tsCommit] = dev
    ? ['devRev', 'devBranch', 'devRepo', Math.floor(Date.now() / 1000)]
    : await Promise.all([
        gitCurrentCommitSha(),
        gitCurrentBranchName(),
        gitCurrentRepoName(),
        gitCurrentCommitTimestamp(),
      ])

  const now = new Date()
  const ts = Math.floor(now.getTime() / 1000)
  const year = String(now.getUTCFullYear())
  const month = String(now.getUTCMonth() + 1).padStart(2, '0')
  const day = String(now.getUTCDate()).padStart(2, '0')
  const hour = String(now.getUTCHours()).padStart(2, '0')
  const minute = String(now.getUTCMinutes()).padStart(2, '0')
  const second = String(now.getUTCSeconds()).padStart(2, '0')

  const tsStr = [[year, month, day].join('-'), [hour, minute, second].join(':')].join(' ')

  const ver = [
    [year, month, day].join(''),
    [hour, minute].join(''),
    repoName,
    branchName,
    rev,
  ].join('_')

  return {
    ts,
    tsCommit,
    tsStr,
    repoName,
    branchName,
    rev,
    ver,
  }
}
