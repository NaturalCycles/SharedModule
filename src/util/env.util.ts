import c from 'chalk'
import * as fs from 'fs-extra'

export interface Json2EnvOpts {
  jsonPath: string
  prefix?: string

  /**
   * @default true
   */
  saveEnvFile?: boolean

  /**
   * @default true
   */
  bashEnv?: boolean

  /**
   * @default true
   */
  fail?: boolean

  debug?: boolean
  silent?: boolean
}

const JSON2ENV_OPT_DEF: Partial<Json2EnvOpts> = {
  saveEnvFile: true,
  bashEnv: true,
  fail: true,
}

export async function json2env (opt: Json2EnvOpts): Promise<void> {
  const { jsonPath, prefix, saveEnvFile, bashEnv, fail, debug, silent } = {
    ...JSON2ENV_OPT_DEF,
    ...opt,
  }

  if (!(await fs.pathExists(jsonPath))) {
    if (fail) {
      throw new Error(`Path doesn't exist: ${jsonPath}`)
    }

    if (!silent) {
      console.log(`json2env input file doesn't exist, skipping without error (${jsonPath})`)
    }

    if (bashEnv) {
      await appendBashEnv('')
    }

    return
  }

  // read file
  const json = await fs.readJson(jsonPath)

  const exportStr = objectToShellExport(json, prefix)
  if (debug) {
    console.log(json, exportStr)
  }

  if (saveEnvFile) {
    const shPath = `${jsonPath}.sh`
    await fs.writeFile(shPath, exportStr)

    if (!silent) {
      console.log(`json2env created ${c.dim(shPath)}:`)
      console.log(exportStr)
    }
  }

  if (bashEnv) {
    await appendBashEnv(exportStr)
  }
}

async function appendBashEnv (exportStr: string): Promise<void> {
  const { BASH_ENV } = process.env
  if (BASH_ENV) {
    await fs.appendFile(BASH_ENV, exportStr + '\n')

    console.log(`BASH_ENV file appended (${c.dim(BASH_ENV)})`)
  }
}

/**
 * Turns Object with keys/values into a *.sh script that exports all keys as values.
 *
 * @example
 * { a: 'b', b: 'c'}
 *
 * will turn into:
 *
 * export a="b"
 * export b="c"
 */
export function objectToShellExport (o: any, prefix = ''): string {
  return Object.keys(o)
    .map(k => {
      const v = o[k]
      if (v) {
        return `export ${prefix}${k}="${v}"`
      }
    })
    .filter(Boolean)
    .join('\n')
}
