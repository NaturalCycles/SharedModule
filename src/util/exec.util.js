const { spawn } = require('child_process')

module.exports.execCommand = async (cmd, exitOnError = true) => {
  return new Promise((resolve, reject) => {
    console.log(cmd)

    const cp = spawn(cmd, {shell: true, stdio: 'inherit'})
    // cp.stdout.on('data', data => console.log(data.toString()))
    // cp.stderr.on('data', data => console.log(data.toString()))
    cp.once('error', err => reject(err))
    cp.once('close', code => {
      // console.log('close: ' + code)
      if (code) {
        if (exitOnError) {
          process.exit(code)
        } else {
          reject(new Error(`${cmd} exitCode: ${code}`))
        }
      } else {
        resolve(code)
      }
    })
  })
}
