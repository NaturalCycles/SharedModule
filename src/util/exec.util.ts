import * as c from 'ansi-colors'
import { SpawnOptions } from 'child_process'
import * as execa from 'execa'

interface ExecOptions extends SpawnOptions {}

export async function proxyCommand (
  cmd: string,
  args: string[] = [],
  opt: ExecOptions = {},
): Promise<void> {
  const [, , ...processArgs] = process.argv

  await execCommand(cmd, [...args, ...processArgs], {
    // shell: true,
    ...opt,
  })
}

export async function execCommand (
  cmd: string,
  args: string[] = [],
  opt: ExecOptions = {},
): Promise<void> {
  const cmdline = [
    ...Object.entries(opt.env || {}).map(([k, v]) => [k, v].join('=')),
    cmd,
    ...args,
  ].join(' ')

  console.log(c.grey(cmdline))

  await execa(cmd, args, {
    stdio: 'inherit',
    env: opt.env,
  }).catch(err => {
    process.exit(err.code || 1)
    throw err // just for tests
  })
}
