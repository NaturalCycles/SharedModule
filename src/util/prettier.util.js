
const { prettierPaths, tslintPaths, tslintExcludePaths } = require('../cnst/prettier.cnst')
const { execCommand } = require('../util/exec.util')

module.exports.runPrettier = async () => {
  // prettier --write 'src/**/*.{js,ts,css,scss,graphql}'
  const cmd = [
    `prettier --write`,
    ...prettierPaths.map(p => `'${p}'`),
  ].join(' ')

  // console.log(cmd)

  return execCommand(cmd)
}

module.exports.runTSLint = async () => {
  // tslint './src/**/*.ts' -e './src/@linked' -t stylish --fix
  const cmd = [
    `tslint`,
    ...tslintPaths.map(p => `'${p}'`),
    ...tslintExcludePaths.map(p => `-e '${p}'`),
    `-t stylish --fix`,
  ].join(' ')

  // console.log(cmd)

  return execCommand(cmd)
}
