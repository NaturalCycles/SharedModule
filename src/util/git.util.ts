import * as execa from 'execa'
import { logExec } from './exec.util'

export async function getLastGitCommitMsg (): Promise<string> {
  // git log -1 --pretty=%B
  const cmd = 'git'
  const args = ['log', '-1', '--pretty=%B']

  const { stdout: msg } = await execa(cmd, args)

  return msg
}

export function commitMessageToTitleMessage (msg: string): string {
  const firstLine = msg.split('\n')[0]
  const [preTitle, title] = firstLine.split(': ')
  return title || preTitle
}

export async function gitHasUncommittedChanges (): Promise<boolean> {
  // git diff-index --quiet HEAD -- || echo "untracked"
  const cmd = 'git diff-index --quiet HEAD --'
  const {code} = await execa(cmd, {
    shell: true,
    reject: false,
  })
  // console.log(code)
  return !!code
}

/**
 * @returns true if there were changes
 */
export async function gitCommitAll (msg: string): Promise<boolean> {
  // git commit -a -m "style(lint-all): $GIT_MSG" || true
  // const cmd = `git commit -a --no-verify -m "${msg}"`
  const cmd = `git`
  const args = [
    'commit',
    '-a',
    '--no-verify',
    '-m',
    msg,
  ]

  logExec(cmd, args)
  const {code} = await execa(cmd, args, {
    // shell: true,
    stdio: 'inherit',
    reject: false,
  })
  console.log(`gitCommitAll code: ${code}`)

  return !code
}

/**
 * @returns true if there are not pushed commits.
 */
export async function gitIsAhead (): Promise<boolean> {
  // ahead=`git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`
  const cmd = `git rev-list HEAD --not --remotes | wc -l | awk '{print $1}'`
  const { stdout } = await execa(cmd, {shell: true})
  console.log(`gitIsAhead: ${stdout}`)
  return Number(stdout) > 0
}

export async function gitPush (): Promise<void> {
  // git push --set-upstream origin $CIRCLE_BRANCH && echo "pushed, exiting" && exit 0
  const cmd = 'git'
  const args = [
    'push',
  ]


  const { CIRCLE_BRANCH } = process.env
  const branchName = CIRCLE_BRANCH || await gitCurrentBranchName()

  if (branchName) {
    args.push('--set-upstream', 'origin', branchName)
  }

  logExec(cmd, args)
  await execa(cmd, args, {
    stdio: 'inherit',
  })
}

export async function gitCurrentBranchName (): Promise<string> {
  // git rev-parse --abbrev-ref HEAD
  const cmd = 'git'
  const args = [
    'rev-parse',
    '--abbrev-ref',
    'HEAD',
  ]

  const { stdout: branchName } = await execa(cmd, args)
  // console.log(`gitCurrentBranchName: ${branchName}`)
  return branchName
}
