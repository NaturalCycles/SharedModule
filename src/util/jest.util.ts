import * as fs from 'fs-extra'
import { cfgDir } from '../cnst/paths.cnst'
import { proxyCommand } from './exec.util'
import { getFullICUPathIfExists } from './test.util'

export function getJestConfigPath (): string | undefined {
  return fs.pathExistsSync(`./jest.config.js`) ? undefined : `${cfgDir}/jest.config.js`
}

export function getJestIntegrationConfigPath (): string | undefined {
  return fs.pathExistsSync(`./jest.integration-test.config.js`)
    ? `./jest.integration-test.config.js`
    : `${cfgDir}/jest.integration-test.config.js`
}

/**
 * Detects if jest is run with all tests, or with specific tests.
 */
export function isRunningAllTests (): boolean {
  const [, , ...args] = process.argv
  const positionalArgs = args.filter(a => !a.startsWith('-'))

  // console.log(process.argv, positionalArgs)

  return !positionalArgs.length
}

/**
 * 1. Detects `full-icu` support, sets NODE_ICU_DATA if needed.
 * 2. Adds `--silent` if running all tests at once.
 */
export async function runJest (ci = false, integration = false): Promise<void> {
  const fullICUPath = getFullICUPathIfExists()
  const jestConfig = integration ? getJestIntegrationConfigPath() : getJestConfigPath()

  const args: string[] = []
  const env = {}

  if (ci) {
    args.push('--ci', '--coverage', '--maxWorkers=2')
  }

  // Running all tests - will use `--silent` to suppress console-logs, will also set process.env.JEST_SILENT=1
  if (ci || isRunningAllTests()) {
    Object.assign(env, {
      JEST_SILENT: '1',
    })

    args.push('--silent')
  }

  if (fullICUPath) {
    Object.assign(env, {
      NODE_ICU_DATA: fullICUPath,
    })
  }

  if (jestConfig) {
    args.push(`--config=${jestConfig}`)
  }

  await proxyCommand('jest', args, {
    env,
  })
}
