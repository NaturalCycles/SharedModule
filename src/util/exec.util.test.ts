import { execCommand } from './exec.util'

test('execCommand ok', async () => {
  const r = await execCommand('ls > /dev/null')
  expect(r).toBe(0)
})

test('execCommand error, exit', async () => {
  process.exit = jest.fn() as any
  await expect(execCommand('abcd')).rejects.toThrow()
  expect(process.exit).toMatchSnapshot()
})

test('execCommand error, reject', async () => {
  await expect(execCommand('abcd', false)).rejects.toThrow()
})
