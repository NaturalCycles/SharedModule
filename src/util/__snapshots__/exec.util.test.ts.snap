// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execCommand error, exit 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      127,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
